lib/authzHttpClient.js

```diff
@@ -0,0 +1,64 @@
+/*!
+ * Copyright (c) 2021 Digital Bazaar, Inc. All rights reserved.
+ */
+import {httpClient} from '@digitalbazaar/http-client';
+import {isInvalidAccessTokenError} from './errorCheck.js';
+import {getAccessToken, getNewAccessToken} from './oauth2-client.js';
+
+const proxyMethods = new Set([
+  'get', 'post', 'push', 'patch', 'head', 'delete'
+]);
+
+export function createAuthzHttpClient({oAuth2Client} = {}) {
+  const authzHttpClient = new Proxy(httpClient, {
+    async apply(target, thisArg, args) {
+      return _handleRequest(target, this, args, oAuth2Client);
+    },
+    get(target, propKey) {
+      const propValue = target[propKey];
+
+      // only intercept particular methods
+      if(!proxyMethods.has(propKey)) {
+        return propValue;
+      }
+      return async function() {
+        return _handleRequest(propValue, this, arguments, oAuth2Client);
+      };
+    }
+  });
+  return authzHttpClient;
+}
+
+async function _handleRequest(target, thisArg, args, oAuth2Client) {
+  const accessToken = await getAccessToken({...oAuth2Client});
+
+  // create the authz header using the token
+  // add the authz header to the headers object of args
+  const [url, options] = args;
+  options.headers = options.headers || {};
+  if(options.headers.Authorization) {
+    options.headers.Authorization = [options.headers.Authorization];
+    options.headers.Authorization.push(`Bearer ${accessToken}`);
+  } else {
+    options.headers.Authorization = `Bearer ${accessToken}`;
+  }
+  try {
+    return target.apply(thisArg, [url, options]);
```

Comment:
If `target.apply` returns a promise then it won't settle within the `try/catch` so `_handleTokenError` won't happen. We need to await it:

```suggestion
    return await target.apply(thisArg, [url, options]);
```

I think if we don't have a test catching this problem we should add one.

Reply:
@dlongley How am I able to test this? In order to get this far I would need to have valid oauth2 credentials, and I don't think we have generic testing ones.

Reply:
Great question! Probably some mock endpoints that expect certain mock tokens would make it easier to test. cc: @aljones15, @dmitrizagidulin -- suggestions?

Reply:
Added 2 tests- one for success and one for failure with an incorrect access token.